### ISSUE TITLE
child process does not receive process signals

### ISSUE THREAD

**Issue by michaelshobbs** [gliderlabs/herokuish#64]
I have (as I assume many do) the need to catch `SIGTERM` upon container shutdown to handle some cleanup. Currently, it seems, herokuish does not pass this through. See below:

example app code ([repo](https://github.com/michaelshobbs/node-js-sample)):
```
var express = require('express');
var app = express();

app.set('port', (process.env.PORT || 5000));
app.use(express.static(__dirname + '/public'));

app.get('/', function(request, response) {
  response.send('Hello World!');
});

process.on( "SIGINT", function() {
  console.log('CLOSING [SIGINT]');
  process.exit();
} );

process.on( "SIGTERM", function() {
  console.log('CLOSING [SIGTERM]');
  process.exit();
} );

app.listen(app.get('port'), function() {
  console.log("Node app is running at localhost:" + app.get('port'));
});
```

Dockerfile.herokuish:
```
FROM gliderlabs/herokuish:latest
EXPOSE 5000

ADD . /app
WORKDIR /app

RUN herokuish buildpack build
CMD herokuish procfile start web
```

Procfile
```
web: node index.js
```

docker
```
$ docker run -d node-js-app setuidgid u17287 /bin/bash -c "PATH=/app/.heroku/node/bin:$PATH node index.js"
4690edb831d53225e1886ce11764176edfd6788a7c98c3a768707fc5a2f28d82
$ docker ps
CONTAINER ID        IMAGE               COMMAND                CREATED              STATUS              PORTS               NAMES
4690edb831d5        node-js-app         "setuidgid u17287 /b   5 seconds ago        Up 4 seconds        5000/tcp            fervent_lovelace
f477a105c114        node-js-app         "/bin/sh -c 'herokui   About a minute ago   Up About a minute   5000/tcp            serene_goldstine
$ docker stop fervent_lovelace
fervent_lovelace
$ docker logs fervent_lovelace
Node app is running at localhost:5000
CLOSING [SIGTERM]
```

herokuish
```
$ docker run -d node-js-app
2b671f27cd1f9568310d2440ce1b334bf5eb6c1e17031cc0ff182d3f49444d40

$ docker ps
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS              PORTS               NAMES
2b671f27cd1f        node-js-app         "/bin/sh -c 'herokui   5 seconds ago       Up 4 seconds        5000/tcp            admiring_heisenberg

$ docker logs admiring_heisenberg
Node app is running at localhost:5000
```

**Comment by michaelshobbs** [comment]
Verified this works on heroku:

```
$ heroku apps:create
Creating someapp... done, stack is cedar-14
https://someapp.herokuapp.com/ | https://git.heroku.com/someapp.git
Git remote heroku added

$ git push heroku master
Counting objects: 439, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (353/353), done.
Writing objects: 100% (439/439), 216.89 KiB | 0 bytes/s, done.
Total 439 (delta 75), reused 391 (delta 51)
remote: Compressing source files... done.
remote: Building source:
remote:
remote: -----> Node.js app detected
remote:
remote: -----> Creating runtime environment
...
remote: Verifying deploy.... done.
To https://git.heroku.com/someapp.git
 * [new branch]      master -> master

$ heroku ps:stop web
Stopping web dynos... done

$ heroku logs
...
2015-08-25T18:26:01.163314+00:00 heroku[slug-compiler]: Slug compilation started
2015-08-25T18:26:01.163338+00:00 heroku[slug-compiler]: Slug compilation finished
2015-08-25T18:26:03.821002+00:00 heroku[web.1]: Starting process with command `node index.js`
2015-08-25T18:26:07.153449+00:00 app[web.1]: Node app is running at localhost:18855
2015-08-25T18:26:07.509882+00:00 heroku[web.1]: State changed from starting to up
2015-08-25T18:27:31.535268+00:00 heroku[web.1]: State changed from up to down
2015-08-25T18:27:34.453910+00:00 heroku[web.1]: Stopping all processes with SIGTERM
2015-08-25T18:27:35.132954+00:00 app[web.1]: CLOSING [SIGTERM]
2015-08-25T18:27:35.969625+00:00 heroku[web.1]: Process exited with status 0
```

**Comment by progrium** [comment]
Ah, yes. Great find. Curious if buildstep did this? I thought it ran Bash, which AFAIK doesn't proxy signals. We'd have that problem here as well, but I think some `exec`'s will fix it. The outer Go program can proxy signals pretty easily, which we can add to https://github.com/progrium/go-basher 

... then we'd probably want an exec here:
https://github.com/gliderlabs/herokuish/blob/master/include/herokuish.bash#L62-Lundefined

... and then another inside the `bash -c` here:
https://github.com/gliderlabs/herokuish/blob/master/include/procfile.bash#L21

This means the processes are exec'd into instead of being subprocesses that would need to proxy signals. The only place we'd need to proxy signals is in Go, which is much easier than Bash.

**Comment by michaelshobbs** [comment]
Oh yeah not a completely new issue. Definitely happened in buildstep too. 

Dockerfile.buildstep
```
FROM progrium/buildstep:latest
EXPOSE 5000

ADD . /app
WORKDIR /app

RUN /build/builder
CMD /start web
```

buildstep:
```
$ docker run -d node-js-app-buildstep
bcc05a9f5b20b6dc95bcaf724a2b8f9803c051ccc22e1edc8a8387a5da502fa9
$ docker ps
CONTAINER ID        IMAGE                   COMMAND                CREATED             STATUS              PORTS               NAMES
bcc05a9f5b20        node-js-app-buildstep   "/bin/sh -c '/start    15 seconds ago      Up 15 seconds       5000/tcp            sleepy_pare
$ docker stop sleepy_pare
docker logs sleepy_pare
sleepy_pare
$ docker logs sleepy_pare
Node app is running at localhost:5000
```

docker (for completeness)
```
$ docker run -d node-js-app-buildstep setuidgid u32535 /bin/bash -c "PATH=/app/.heroku/node/bin:$PATH node index.js"
d435d7cc494798056f8a63d7c05a31981c3d08dabdae735c9398f9556ade03a0
$ docker ps
CONTAINER ID        IMAGE                   COMMAND                CREATED             STATUS              PORTS               NAMES
d435d7cc4947        node-js-app-buildstep   "setuidgid u32535 /b   23 seconds ago      Up 23 seconds       5000/tcp            trusting_hoover
$ docker stop trusting_hoover
trusting_hoover
$ docker logs trusting_hoover
Node app is running at localhost:5000
CLOSING [SIGTERM]
```

**Comment by michaelshobbs** [comment]
@progrium I can definitely make a PR against herokuish easily. Not too sure about go-basher.

**Comment by progrium** [comment]
If you just make an issue in go-basher to track it I can be sure to get it
done myself or by GL staff. ;)

On Tue, Aug 25, 2015 at 2:07 PM, Michael Hobbs <notifications@github.com>
wrote:

> @progrium <https://github.com/progrium> I can definitely make a PR
> against herokuish easily. Not too sure about go-basher.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gliderlabs/herokuish/issues/64#issuecomment-134704086>
> .
>



-- 
Jeff Lindsay
http://progrium.com

**Comment by michaelshobbs** [comment]
@progrium in herokuish.bash i'm not sure where the exec should go. something like this?

```
	exec setuidgid "$unprivileged_user" "$@"
```

**Comment by progrium** [comment]
Yeah, that would turn the current Bash process started by go-basher that's
running all this bash into setuidgid, which presumably execs into the
second argument.

However, since we eval the Procfile line and run in Bash, the way this is
called in procfile-exec is just execing into Bash again, so we'd need an
exec inside that one as well.

On Tue, Aug 25, 2015 at 2:42 PM, Michael Hobbs <notifications@github.com>
wrote:

> @progrium <https://github.com/progrium> in herokuish.bash i'm not sure
> where the exec should go. something like this?
>
>     exec setuidgid "$unprivileged_user" "$@"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gliderlabs/herokuish/issues/64#issuecomment-134714222>
> .
>



-- 
Jeff Lindsay
http://progrium.com

**Comment by josegonzalez** [comment]
@michaelshobbs is everything here fixed?

**Comment by michaelshobbs** [comment]
Fixed by #65
